### PR TITLE
Optional k8s charm pv

### PR DESCRIPTION
**Pull request by wallyworld**
### Checklist

 - [x] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?

There's a change to api parameters but it's for a controller facade.

----

## Description of change

A k8s charm can now signal that it does not require a PV to be created for charm state storage. It does this by setting min-juju-version=2.8.0. Existing charms which do require PV storage are still catered for, and if a charm is upgraded, any PV is retained.

The first commit factors out the min version check to a common function. It has been tweaked to allow a charm specifying a major.minor.0 version to work with a beta/rc (to allow testing).

## QA steps

Deploy a k8s charm
Edit the charm to add min-juju-version=2.8.0
upgrade the charm and check the PV and PVC are still used
deploy the edited charm again as a different application
check that no operator PV is created

### CODE REVIEW

File: caas/kubernetes/provider/operator.go [hunk 1]
```diff
@@ -197,44 +197,7 @@ func (k *kubernetesClient) EnsureOperator(appName, agentPath string, config *caa
 		}
 	}
 
-	// Set up the parameters for creating charm storage.
```
Comment (wallyworld) [inline caas/kubernetes/provider/operator.go]: This code is moved to a new separate operatorVolumeClaim() function below

**Comment by wallyworld** [comment]
[k8s charms do not require storage](https://trello.com/c/vhhzwBXc/1079-k8s-charms-do-not-require-storage)

File: caas/broker.go [hunk 2]
```diff
@@ -371,9 +371,9 @@ type OperatorConfig struct {
 	// Version is the Juju version of the operator image.
 	Version version.Number
 
-	// CharmStorage defines parameters used to create storage
-	// for operators to use for charm state.
-	CharmStorage CharmStorageParams
+	// CharmStorage defines parameters used to optionally
+	//create storage for operators to use for charm state.
```
Comment (hpidcock) [inline caas/broker.go]: //(space)create storage

File: caas/kubernetes/provider/providerconfig.go [hunk 3]
```diff
@@ -7,16 +7,34 @@ import (
...
 	"github.com/juju/schema"
+	"github.com/juju/version"
 	"gopkg.in/juju/environschema.v1"
 
 	"github.com/juju/juju/environs/config"
 )
 
 const (
+	// WorkloadStorageKey is the model config attribute used to specify
+	//the storage class for provisioning workload storage.
```
Comment (hpidcock) [inline caas/kubernetes/provider/providerconfig.go]: I need some space here

File: caas/kubernetes/provider/providerconfig.go [hunk 4]
```diff
@@ -7,16 +7,34 @@ import (
...
 	WorkloadStorageKey = "workload-storage"
+
+	// OperatorStorageKey is the model config attribute used to specify
+	//the storage class for provisioning operator storage.
```
Comment (hpidcock) [inline caas/kubernetes/provider/providerconfig.go]: and here

File: caas/kubernetes/provider/operator.go [hunk 5]
```diff
@@ -269,20 +231,80 @@ func (k *kubernetesClient) EnsureOperator(appName, agentPath string, config *caa
...
+func (k *kubernetesClient) operatorVolumeClaim(appName, operatorName string, storageParams *caas.CharmStorageParams) (*core.PersistentVolumeClaim, error) {
+	// We may no longer need storage for charms, but if the charm has previously been deployed
+	// with storage, we need to retain that.
+	operatorVolumeClaim := "charm"
+	if isLegacyName(operatorName) {
+		operatorVolumeClaim = fmt.Sprintf("%v-operator-volume", appName)
+	}
+	if storageParams == nil {
+		existingClaim, err := k.getPVC(operatorVolumeClaim)
+		if err != nil && !errors.IsNotFound(err) {
```
Comment (hpidcock) [inline caas/kubernetes/provider/operator.go]: Ugh I'd prefer this
```
if errors.IsNotFound(err) {
    logger.Debugf("found existing volume claim for operator %s", operatorName)
    return nil, nil
} else if err != nil {
    return nil, errors.Annotatef(err, "getting operator volume claim")
}
return existingClaim, nil
```

File: caas/kubernetes/provider/operator_test.go [hunk 6]
```diff
@@ -621,6 +621,138 @@ func (s *K8sBrokerSuite) TestEnsureOperatorUpdate(c *gc.C) {
 	}
 }
 
+func (s *K8sBrokerSuite) TestEnsureOperatorNoStorage(c *gc.C) {
```
Comment (hpidcock) [inline caas/kubernetes/provider/operator_test.go]: This test is TestEnsureOperatorNoStorageKeepsPVC

Need another to ensure it handles no existing PVC and doesn't create one.

Comment (wallyworld) [reply caas/kubernetes/provider/operator_test.go]: yeah, just added a test for that. without a pvc, gets deployed as a deployment

**Comment by wallyworld** [comment]
$$merge$$

**Comment by hpidcock** [comment]
I don't think we should be using deployments for the operator. Statefulsets have guarantees with regards to the number of pods running. I don't believe the operator/uniter logic will play well at all with two pods running the operator.

**Pull request merged**